### PR TITLE
More specs to increase test coverage

### PR DESCRIPTION
**Pull request by antondalgren**

### CODE REVIEW

File: test/amqp-view-test.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import AMQPView from '../src/amqp-view.js';
+
+test("setFrameEnd", async t => {
+  const view = new AMQPView(new ArrayBuffer(30))
+  t.is(view.setFrameEnd(0, true), 1)
+})
+
+test("setField with ArrayBuffer",  async t => {
+  const view = new AMQPView(new ArrayBuffer(30))
+  t.is(view.setField(0, new ArrayBuffer(1)), 6)
```
Comment (carlhoerberg) [inline test/amqp-view-test.ts:11]: can add this to the "can publish all types of properties" test instead

File: test/amqp-view-test.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+})
+
+test("setField with bigint", async t => {
```
Comment (carlhoerberg) [inline test/amqp-view-test.ts:14]: same with this one

File: test/amqp-view-test.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import test from 'ava';
+import AMQPView from '../src/amqp-view.js';
+
+test("setFrameEnd", async t => {
```
Comment (carlhoerberg) [inline test/amqp-view-test.ts:4]: we could just remove the method instead, as it's not used

File: test/test.ts [hunk 4]
```diff
@@ -427,3 +427,70 @@ test('can open a specific channel twice', async t => {
   const ch2 = await conn.channel(2)
   t.is(ch2, ch)
 })
+
+
+test('set basic flow on channel', async t => {
+  const amqp = new AMQPClient("amqp://127.0.0.1")
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  await t.notThrowsAsync(async () =>  await ch.basicFlow(true))
```
Comment (carlhoerberg) [inline test/test.ts:436]: doesn't this throw with rabbitmq? or is it just `basicFlow(false)` that does?

File: test/test.ts [hunk 5]
```diff
@@ -427,3 +427,70 @@ test('can open a specific channel twice', async t => {
...
+})
+
+test('can resolve promise on channel', async t => {
+  const amqp = new AMQPClient("amqp://127.0.0.1")
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.false(ch.resolvePromise())
```
Comment (carlhoerberg) [inline test/test.ts:443]: resolvePromise should be a private field, what does this really test?

File: test/test.ts [hunk 6]
```diff
@@ -427,3 +427,70 @@ test('can open a specific channel twice', async t => {
...
+})
+
+test('confirming unknown deliveryTag', async t => {
+  const amqp = new AMQPClient("amqp://127.0.0.1")
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.publishConfirmed(1, false, false))
```
Comment (carlhoerberg) [inline test/test.ts:450]: this doesn't make sense?? publishConfirm is an internal method. can't we excerise the publish confirm code with a realistic test?

File: test/test.ts [hunk 7]
```diff
@@ -427,3 +427,70 @@ test('can open a specific channel twice', async t => {
...
+  const msg = new AMQPMessage(ch)
+  msg.consumerTag = "abc"
+  t.throws(() => ch.deliver(msg))
+})
+
+test("can publish as json", async t => {
+  const amqp = new AMQPClient("amqp://127.0.0.1")
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.basicPublish("amq.topic", "", {test: "hey"}))
```
Comment (carlhoerberg) [inline test/test.ts:467]: myabe we shouldn't support this? amqplib doesn't..

File: test/test.ts [hunk 8]
```diff
@@ -427,3 +427,70 @@ test('can open a specific channel twice', async t => {
...
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.basicPublish("amq.topic", "", null))
+})
+
+test("can publish ArrayBuffer", async t => {
+  const amqp = new AMQPClient("amqp://127.0.0.1")
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.basicPublish("amq.topic", "", new ArrayBuffer(2)))
```
Comment (carlhoerberg) [inline test/test.ts:481]: should try to get the message too, to verify that we get the same data as we published.

File: test/test.ts [hunk 9]
```diff
@@ -427,3 +427,70 @@ test('can open a specific channel twice', async t => {
...
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.basicPublish("amq.topic", "", {test: "hey"}))
+})
+
+test("can publish null", async t => {
+  const amqp = new AMQPClient("amqp://127.0.0.1")
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.basicPublish("amq.topic", "", null))
```
Comment (carlhoerberg) [inline test/test.ts:474]: should try to get the message too, to verify that we get the same data as we published.

File: test/test.ts [hunk 10]
```diff
@@ -427,3 +427,70 @@ test('can open a specific channel twice', async t => {
...
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.basicPublish("amq.topic", "", new ArrayBuffer(2)))
+})
+
+test("can publish Uint8array", async t => {
+  const amqp = new AMQPClient("amqp://127.0.0.1")
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.basicPublish("amq.topic", "", new Uint8Array(2)))
```
Comment (carlhoerberg) [inline test/test.ts:488]: should try to get the message too, to verify that we get the same data as we published.

File: test/test.ts [hunk 11]
```diff
@@ -427,3 +427,70 @@ test('can open a specific channel twice', async t => {
...
+})
+
+test("can do basicRecover", async t => {
+  const amqp = new AMQPClient("amqp://127.0.0.1")
+  const conn = await amqp.connect()
+  const ch = await conn.channel()
+  t.notThrows(() => ch.basicRecover(true))
```
Comment (carlhoerberg) [inline test/test.ts:495]: maybe a real test? get a message with noAck false, then basicRecover(false), then try to ack the message, that should throw

File: src/amqp-channel.ts [hunk 12]
```diff
@@ -274,12 +274,14 @@ export default class AMQPChannel {
    * @param [immediate] - if the message should be returned if it can't be delivered to a consumer immediately (not supported in RabbitMQ)
    * @return - fulfilled when the message is enqueue on the socket, or if publish confirm is enabled when the message is confirmed by the server
    */
-  basicPublish(exchange: string, routingKey: string, data: string|Uint8Array|ArrayBuffer|null, properties: AMQPProperties = {}, mandatory = false, immediate = false) {
+  basicPublish(exchange: string, routingKey: string,
+               data: string|Uint8Array|ArrayBuffer|null|Record<string|symbol|number, unknown>,
```
Comment (carlhoerberg) [inline src/amqp-channel.ts:278]: i think it's a mistake to allow "objects", and that the lib autoamtically encodes it as json.

File: test/test.ts (hunk 4 above)
Comment (antondalgren) [reply test/test.ts:436]: `basicFlow(false)` throws.

File: test/test.ts (hunk 5 above)
Comment (antondalgren) [reply test/test.ts:443]: That the method returns false if there are no promises to resolve.

File: test/test.ts (hunk 6 above)
Comment (antondalgren) [reply test/test.ts:450]: The happy path is already covered with a "realistic test". If it doesn't make sense to test what happens if we try to confirm a unknown deliverytag, what's the reason for having that code path?

File: test/test.ts (hunk 9 above)
Comment (antondalgren) [reply test/test.ts:474]: I don't think that's necessary as we have that covered in other tests. But I'll add it!

File: test/test.ts (hunk 6 above)
Comment (carlhoerberg) [reply test/test.ts:450]: Ok, alright, but maybe we should do something more than just not throw. Do we log? Should we have an on('error') emitter?

File: test/test.ts (hunk 11 above)
Comment (antondalgren) [reply test/test.ts:488]: Can't get this to work... `msg.ack()` does not throw the error, but rather a reject established with the connection. I don't know how to handle that.

File: test/test.ts (hunk 9 above)
Comment (antondalgren) [reply test/test.ts:467]: I'm not managing to do that, not sure what I'm doing wrong... 🤷

**Pull request merged**